bin/extract-firmware

```diff
@@ -0,0 +1,389 @@
+#!/usr/bin/perl
+# Extract all of the files from a UEFI firmware image.
+#
+# This is a simple replacement for uefi-firmware-parser
+# with fewer features and more suited for reconstituting the
+# firmware later.
+#
+use warnings;
+use strict;
+use FindBin;
+use lib "$FindBin::Bin/../lib";
+use EFI;
+use Getopt::Long;
+use File::Basename;
+use File::Temp 'tempfile';
+use Data::Dumper;
+
+my $usage = <<"";
+Usage:
+  $0 [options] firmware.rom | tee firmware.txt
+Options:
+  -v | --version          Print per-section logs
+  -o | --output-dir DIR   Write the output to a different base directory
+  -s | --start 0xX        Starting offset for partial extraction
+  -e | --end 0xX          Ending offset
+
+my $start_offset = 0; # 0x02c00000;
+my $end_offset;
+my $verbose = 0;
+my $base_dir = '.';
+
+GetOptions(
+	"v|verbose+"		=> \$verbose,
+	"o|output-dir=s"	=> \$base_dir,
+	"s|start=o"		=> \$start_offset,
+	"e|end=o"		=> \$end_offset,
+) or die $usage;
+
+local $/;
+while(<>)
+{
+	process_region($base_dir, $_);
+}
+
+sub process_region
+{
+	my $base = shift;
+	my $data = shift;
+	my $length = length($data);
+
+	printf "%s: length 0x%x\n", $base, $length
+		if $verbose;
+
+	my $start_unknown;
+
+	# Search for the start of firmware volumes,
+	# identified by their '_FVH' in the structure
+	my $step = 256;
+
+	# Adjust the end offset if they ask for too much
+	$end_offset = $length - 0x30
+		if !defined $end_offset or $end_offset > $length - 0x30;
+
+	for(my $offset = $start_offset ; $offset < $end_offset ; $offset += $step)
+	{
+		# Look for a flash region descriptor
+		my $ifd_sig = unpack("N", substr($_, $offset + 0x10, 4));
+		if ($ifd_sig == 0x5AA5F00F)
+		{
+			printf "%s/0x%08x.ifd: Flash Descriptor\n",
+				$base,
+				$offset,
+				;
+
+			my $ifd_len = 0xF0000;
```

Comment:
That was a typo.  I changed it to 0x10000, which seems to be the size on most of the platforms.  This is to avoids parsing the IFD, especially since we might not have one for some mainboards (like qemu).